sstables/sstable_set.cc

```diff
@@ -184,6 +184,11 @@ void sstable_set::for_each_sstable(std::function<void(const shared_sstable&)> fu
     return _impl->for_each_sstable(std::move(func));
 }
 
+bool
+sstable_set::contains(generation_type gen) const noexcept {
+    return _impl->contains(gen);
+}
+
```

Comment:
There is overlap in the sstable_map and sstable_set APIs indeed, but `sstable_set` is more abstract and higher level.
I don't see the problem in keeping both levels of abstraction and use sstable_map to compose sstable_set implementations.

Reply:
I am more worried about the maintenance work for the future maintainers. In general, we should avoid introducing new mechanisms that aren't really needed to solve a particular problem.

In the commit message of 'sstable_set: turn sstable_list into unordered_map<generation, shared_sstable>' it says:

```
This has the advantage that a sstable can be easily looked
up by its generation (to be used in the next patches
for determining if a staging sstable already has a paired
sstable in the base directory)
```

but with the introduction of sstable_set::contains(generation_type gen), we can do exactly it. there's not even a need to change the signature of table::_sstables_staging, which is an unordered_map. 

sstable_map is a nice utility, but I don't think it's needed for solving this problem. with sstable_set now being able to look up a file through generation_type, we can still rely on old simple sstable_list for basic operations over a sstable list. sstable_list as an unordered_set brings simplicity, which I like, when exposing the sstable set to the outside world via table::get_sstables().

Reply:
Let's separate forward development and maintenance / stability.
They represent conflicting gravity attractors.
On one hand you want to keep things stable and easy to backport.
That nice and defensive, but if it's imbalanced, you end up with stagnation and accumulation of deadweight which is the old infrastructure you kwwp maintaining instead of overhauling it from time to get new, shiny plumbing.  Pretty much what happened with the compaction_manager over time where patches accumulated with the intention to change little to reduce the risk, but eventually this lead to disarray.

The question here is if there's even intention to backport this change, as a fix, or consider it a feature and keep the focus forward-looking.

If it's the first case, I'm concur to implementing the fix with minimal changes first, for easy backporting, and then following up with the deeper cleanup.  But if it's the latter case (which I suspect) then this might not be worth the trouble.

Reply:
I'd suggest introducing sstable_map in a separate series, so the motivation can be explained and discussed in its own series. And we get the benefit of your series being easily backported.

Reply:
WIP in https://github.com/bhalevy/scylla/tree/unique_genration_sstable_set